File: Solutions/0481/0481.c
/***************************************************************************************************
 * Copyright: washing
 * FileName: 0481.c
 * Author: washing
 * Version: 1.0
 * Date: 2022-10-31 07:56:07
***************************************************************************************************/

int magicalString(int n){
    if (n <= 3) return 1;
    int *l = (int *)malloc(sizeof(int) * n);
    l[0] = 1;
    l[1] = 2;
    l[2] = 2;
    int ret = 1;
    int gai = 1;
    int idx1 = 2;
    int idx2 = 3;
    for (int i=2; i<n; i++) {
        for (int j=0; j<l[idx1]; j++) {
            l[idx2++] = gai;
            if (gai == 1) ret++;
            if (idx2 == n) return ret;
        }
        idx1++;
        gai = gai % 2 + 1;
    }
    return 0;
}

File: Solutions/0481/0481.py
# -*- coding:utf-8 -*-
# Author: washing
# DateTime: 2022/10/31 07:56
# File: 0481.py
# Desc: 


class Solution:
    def magicalString(self, n: int) -> int:
        if n <= 3: return 1
        l = [1,2,2]
        gai = 1
        idx = 2
        for _ in range(2, n):
            l.extend([gai] * l[idx])
            idx += 1
            if gai == 1: gai = 2
            else: gai = 1
            if len(l) >= n:
                return sum([i%2 for i in l[:n]])
